soccer/gameplay/plays/offense/adaptive_formation.py

```diff
@@ -0,0 +1,295 @@
+        collecting = 1
+        # Dribble for a second and prepare to pass / shoot / clear
+        dribbling = 2
+        # Pass when someone is open
+        passing = 3
+        # Shoot when chances are high
+        shooting = 4
+        # Clear when pass / dribble is worse and we are in our own zone
+        clearing = 5
+
+    def __init__(self):
+        super().__init__(continuous=False)
+
+        for s in AdaptiveFormation.State:
+            self.add_state(s, behavior.Behavior.State.running)
+
+        # Dribbling skill
+        self.dribbler = None
+        # Dribble start point
+        self.dribble_start_pt = robocup.Point(0, 0)
+
+        # Kicker for a shot
+        self.kick = None
+        # Controls robots while passes are being set up
+        self.midfielders = None
+
+        # State Decision Variables
+        self.shot_chance = 0
+        self.pass_score = 0
+        # Prev State Decision Variables
+        self.prev_shot_chance = 0
+        self.prev_pass_score = 0
+        # Used to keep dribble within rules
+        self.check_dribbling_timer = 0
+        self.check_dribbling_timer_cutoff = 100
+
+        self.kick_eval = robocup.KickEvaluator(main.system_state())
+
+        # Add transitions
+        self.add_transition(behavior.Behavior.State.start,
+                            AdaptiveFormation.State.collecting, lambda: True,
+                            'immediately')
+
+        self.add_transition(
+            AdaptiveFormation.State.collecting,
+            AdaptiveFormation.State.dribbling,
+            lambda: self.subbehavior_with_name('defend').state == behavior.Behavior.State.completed,
+            'Ball Collected')
+
+        self.add_transition(
+            AdaptiveFormation.State.dribbling, AdaptiveFormation.State.passing,
+            lambda: self.should_pass_from_dribble() and not self.should_shoot_from_dribble(),
+            'Passing')
+
+        self.add_transition(AdaptiveFormation.State.dribbling,
+                            AdaptiveFormation.State.shooting,
+                            lambda: self.should_shoot_from_dribble(),
+                            'Shooting')
+
+        self.add_transition(AdaptiveFormation.State.dribbling,
+                            AdaptiveFormation.State.clearing,
+                            lambda: self.should_clear_from_dribble(),
+                            'Clearing')
+
+        self.add_transition(
+            AdaptiveFormation.State.passing, AdaptiveFormation.State.dribbling,
+            lambda: self.subbehavior_with_name('pass').state == behavior.Behavior.State.completed,
+            'Passed')
+
+        # Reset to collecting when ball is lost at any stage
+        self.add_transition(
+            AdaptiveFormation.State.dribbling,
+            AdaptiveFormation.State.collecting,
+            lambda: self.dribbler is not None and self.dribbler.robot is not None and evaluation.ball.robot_has_ball(self.dribbler.robot),
```

Comment:
These transitions `dribbling->{collecting,clearing,shooting}` have the possibility of happening at the same time. Right now it prefers collecting since that's the lowest Enum number, but it would be preferable to explicitly make sure there's only one option at a time. Could you change the logic to depend on calls to the other states (for example, maybe prefer collecting to shooting and clearing).